File: processor/src/test/java/io/michaelrocks/lightsaber/processor/annotations/AnnotationHelper.java
/*
 * Copyright 2015 Michael Rozumyanskiy
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package io.michaelrocks.lightsaber.processor.annotations;

import org.apache.commons.lang3.tuple.Pair;
import org.objectweb.asm.Type;

import java.util.Collections;
import java.util.HashMap;
import java.util.Map;

class AnnotationHelper {
    public static AnnotationDescriptor createAnnotationDescriptor(final String annotationName) {
        return new AnnotationDescriptor(getAnnotationType(annotationName), Collections.<String, Type>emptyMap());
    }

    public static AnnotationDescriptor createAnnotationDescriptor(final String annotationName, final Type type) {
        return createAnnotationDescriptor(annotationName, "value", type);
    }

    public static AnnotationDescriptor createAnnotationDescriptor(final String annotationName, final String name,
            final Type type) {
        return createAnnotationDescriptor(annotationName, Collections.singletonMap(name, type));
    }

    @SafeVarargs
    public static AnnotationDescriptor createAnnotationDescriptor(final String annotationName,
            final Pair<String, Type>... values) {
        return new AnnotationDescriptor(getAnnotationType(annotationName), pairsToMap(values));
    }

    public static AnnotationData createAnnotationData(final String annotationName) {
        return new AnnotationData(getAnnotationType(annotationName), Collections.<String, Object>emptyMap(), false);
    }

    public static AnnotationData createAnnotationData(final String annotationName, final Object defaultValue) {
        return createAnnotationData(annotationName, "value", defaultValue);
    }

    public static AnnotationData createAnnotationData(final String annotationName, final String methodName,
            final Object defaultValue) {
        return createAnnotationData(annotationName, Collections.singletonMap(methodName, defaultValue));
    }

    public static AnnotationDescriptor createAnnotationDescriptor(final String annotationName,
            final Map<String, Type> fields) {
        return new AnnotationDescriptor(getAnnotationType(annotationName), fields);
    }

    @SafeVarargs
    public static AnnotationData createAnnotationData(final String annotationName,
            final Pair<String, ?>... values) {
        return new AnnotationData(getAnnotationType(annotationName), pairsToMap(values), false);
    }

    public static AnnotationData createAnnotationData(final String annotationName,
            final Map<String, Object> values) {
        return new AnnotationData(getAnnotationType(annotationName), values, false);
    }

    public static AnnotationData createResolvedAnnotationData(final String annotationName) {
        return new AnnotationData(getAnnotationType(annotationName), Collections.<String, Object>emptyMap(), true);
    }

    public static AnnotationData createResolvedAnnotationData(final String annotationName, final Object defaultValue) {
        return createResolvedAnnotationData(annotationName, "value", defaultValue);
    }

    public static AnnotationData createResolvedAnnotationData(final String annotationName, final String methodName,
            final Object defaultValue) {
        return createResolvedAnnotationData(annotationName, Collections.singletonMap(methodName, defaultValue));
    }

    @SafeVarargs
    public static AnnotationData createResolvedAnnotationData(final String annotationName,
            final Pair<String, ?>... values) {
        return new AnnotationData(getAnnotationType(annotationName), pairsToMap(values), false);
    }

    public static AnnotationData createResolvedAnnotationData(final String annotationName,
            final Map<String, Object> values) {
        return new AnnotationData(getAnnotationType(annotationName), values, true);
    }

    public static Type getAnnotationType(final String annotationName) {
        return Type.getObjectType(annotationName);
    }

    private static <T> Map<String, T> pairsToMap(final Pair<String, ? extends T>[] pairs) {
        final Map<String, T> values = new HashMap<>();
        for (final Pair<String, ? extends T> pair : pairs) {
            values.put(pair.getKey(), pair.getValue());
        }
        return values;
    }
}

File: processor/src/test/kotlin/io/michaelrocks/lightsaber/processor/annotations/AnnotationHelper.kt
/*
 * Copyright 2015 Michael Rozumyanskiy
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package io.michaelrocks.lightsaber.processor.annotations

import org.objectweb.asm.Type
import java.util.*

internal object AnnotationHelper {
  fun createAnnotationDescriptor(annotationName: String): AnnotationDescriptor =
      AnnotationDescriptor(getAnnotationType(annotationName), emptyMap<String, Type>())

  fun createAnnotationDescriptor(annotationName: String, type: Type): AnnotationDescriptor =
      createAnnotationDescriptor(annotationName, "value", type)

  fun createAnnotationDescriptor(annotationName: String, name: String, type: Type): AnnotationDescriptor =
      createAnnotationDescriptor(annotationName, Collections.singletonMap(name, type))

  fun createAnnotationDescriptor(annotationName: String, vararg values: Pair<String, Type>): AnnotationDescriptor =
      AnnotationDescriptor(getAnnotationType(annotationName), hashMapOf(*values))

  fun createAnnotationDescriptor(annotationName: String, fields: Map<String, Type>): AnnotationDescriptor =
      AnnotationDescriptor(getAnnotationType(annotationName), fields)

  fun createAnnotationData(annotationName: String): AnnotationData =
      AnnotationData(getAnnotationType(annotationName), emptyMap<String, Any>())

  fun createAnnotationData(annotationName: String, defaultValue: Any): AnnotationData =
      createAnnotationData(annotationName, "value", defaultValue)

  fun createAnnotationData(annotationName: String, methodName: String, defaultValue: Any): AnnotationData =
      createAnnotationData(annotationName, Collections.singletonMap(methodName, defaultValue))

  fun createAnnotationData(annotationName: String, vararg values: Pair<String, Any>): AnnotationData =
      AnnotationData(getAnnotationType(annotationName), hashMapOf(*values))

  fun createAnnotationData(annotationName: String, values: Map<String, Any>): AnnotationData =
      AnnotationData(getAnnotationType(annotationName), values)

  fun getAnnotationType(annotationName: String): Type =
      Type.getObjectType(annotationName)
}
